File: src/ARCtrl/WebRequest/WebRequest.fs
module ARCtrl.WebRequest

open Fable.Core

#if !FABLE_COMPILER_PYTHON
open Fable.SimpleHttp
#endif

let downloadFile url =
    #if !FABLE_COMPILER_PYTHON
    let browserAndDotnet() =  
        async {
            let! (statusCode, responseText) = Http.get url

            return 
                match statusCode with
                | 200 -> responseText
                | _ -> failwithf "Status %d => %s" statusCode responseText
        }
    #endif
    #if FABLE_COMPILER_JAVASCRIPT
    
    if ARCtrl.WebRequestHelpers.NodeJs.isNode() then
        // From here: https://github.com/fable-compiler/fable3-samples/blob/25ea2404b28c897988b144f0141bc116da292679/nodejs/src/App.fs#L7
        Fable.Core.JsInterop.importSideEffects "isomorphic-fetch"
        ARCtrl.WebRequestHelpers.NodeJs.downloadFile url
    else
        browserAndDotnet()
    #endif

    #if FABLE_COMPILER_PYTHON
    ARCtrl.WebRequestHelpers.Py.downloadFile url
    #endif

    #if !FABLE_COMPILER
    browserAndDotnet()
    #endif

File: js/WebRequest/WebRequest.js
import { downloadFile as downloadFile_1, isNode } from "./WebRequest.Node.js";
import { singleton } from "../fable_modules/fable-library-js.4.22.0/AsyncBuilder.js";
import { Http_get } from "../fable_modules/Fable.SimpleHttp.3.5.0/Http.fs.js";
import { printf, toFail } from "../fable_modules/fable-library-js.4.22.0/String.js";
import "isomorphic-fetch";

export function downloadFile(url) {
    if (isNode()) {
        return downloadFile_1(url);
    }
    else {
        return singleton.Delay(() => singleton.Bind(Http_get(url), (_arg) => {
            const statusCode = _arg[0] | 0;
            const responseText = _arg[1];
            return singleton.Return((statusCode === 200) ? responseText : toFail(printf("Status %d => %s"))(statusCode)(responseText));
        }));
    }
}

